### PR TITLE
feat(core): internal stack integration with mocha runner

### LINKED ISSUE #838: Problem installing spec reporter with Angular 12 and mocha
Hi there! I am in the process of moving our application from Angular 11 to Angular 12. Since updating, we now get the following error from ng-mocks when I try to run our unit-testing suite:

```
Uncaught Error: ng-mocks cannot install own spec reporter. This affects its core features for MockInstance and MockRender. Please report an issue on github. If you use jest v27, please add to its config testRunner=jest-jasmine2 for now and upvote the issue on github: https://github.com/facebook/jest/issues/11483.
at http://localhost:9876/_karma_webpack_/vendor.js:196649:9

Error: ng-mocks cannot install own spec reporter. This affects its core features for MockInstance and MockRender. Please report an issue on github. If you use jest v27, please add to its config testRunner=jest-jasmine2 for now and upvote the issue on github: https://github.com/facebook/jest/issues/11483.
  at messageCoreChecker (http://localhost:9876/_karma_webpack_/webpack:/node_modules/ng-mocks/cjs/lib/common/ng-mocks-stack.js:119:1)
  at install (http://localhost:9876/_karma_webpack_/webpack:/node_modules/ng-mocks/cjs/lib/common/ng-mocks-stack.js:172:1)
  at Object.23851 (http://localhost:9876/_karma_webpack_/webpack:/node_modules/ng-mocks/cjs/lib/common/ng-mocks-stack.js:178:1)
  at __webpack_require__ (http://localhost:9876/_karma_webpack_/webpack:/webpack/bootstrap:19:1)
  at Object.71794 (http://localhost:9876/_karma_webpack_/webpack:/node_modules/ng-mocks/cjs/index.js:15:1)
  at __webpack_require__ (http://localhost:9876/_karma_webpack_/webpack:/webpack/bootstrap:19:1)
  at Module.4825 (http://localhost:9876/_karma_webpack_/main.js:30055:66)
  at __webpack_require__ (http://localhost:9876/_karma_webpack_/webpack:/webpack/bootstrap:19:1)
  at http://localhost:9876/_karma_webpack_/webpack:/webpack/startup:4:1
  at Function.__webpack_require__.O (http://localhost:9876/_karma_webpack_/webpack:/webpack/runtime/chunk loaded:23:1)
```

Looking through [ng-mocks-stack](https://github.com/ike18t/ng-mocks/blob/71d9686f4669994b314d4199749fd3e3d3fe4856/libs/ng-mocks/src/lib/common/ng-mocks-stack.ts#L141-L156), I suspect this is because we use [mocha](https://github.com/mochajs/mocha) as our test reporter, whereas ng-mocks is trying to install reporters for either Jasmine or Jest.

We are running ng-mocks 12.3.1. Here's the output from ng version:
```
Angular CLI: 12.1.2
Node: 14.15.1
Package Manager: yarn 1.22.5
OS: linux x64

Angular: 12.1.2
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, upgrade

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1201.2
@angular-devkit/build-angular   12.1.2
@angular-devkit/core            12.1.2
@angular-devkit/schematics      12.1.2
@ngtools/webpack                12.1.2
@schematics/angular             12.1.2
rxjs                            6.6.7
typescript                      4.3.5
webpack                         5.42.0
```

### PR DESCRIPTION
**Pull request by dpraul**
Add mocha runner support to the internal stack.

Closes #838

### CODE REVIEW

**Review by satanTime — APPROVED** [review]

**Comment by satanTime** [comment]
Hi @dpraul,

thanks for the PR. Two things:

- could you sign the commit? https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work
- could you add the ticket number to the end of the commit title? `feat(core): internal stack integration with mocha runner #838`

Thank you in advance!

File: libs/ng-mocks/src/lib/common/ng-mocks-stack.ts [hunk 1]
```diff
@@ -138,11 +138,30 @@ const installJestCircus = () => {
   return true;
 };
 
+// istanbul ignore next
+const installMochaReporter = () => {
+  try {
+    mocha.setup({
+      rootHooks: {
+        afterAll: stackPop,
+        beforeAll: () => stackPush('root'),
```
Comment (satanTime) [inline libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: Can it be that beforeAll is used several times per test file?

If it is so, then it should be `suite`.

`root` is provided by `ng-mocks` on the start.

Comment (dpraul) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: According to https://mochajs.org/#available-root-hooks, the `beforeAll`/`afterAll` root hooks behave differently depending on if mocha is run in serial or parallel mode. They seem mostly synonymous to ng-mocks `root`, but in parallel mode mocha has multiple "roots". I believe parallel mode would instantiate a different ng-mocks instance per each parallel run, so it should be fine as `root`, but I cannot verify as we use serial mode.

Comment (satanTime) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: Can you do something like that with mocha:

```
describe
  beforeAll
  afterAll
  it
  it
  it
  describe
    beforeAll
    afterAll
    it
    it
```

so the first beforeAll covers the nested describe too, whereas the second beforeAll covers only its suite.

Comment (dpraul) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: Yes you can, but the hooks here do not cause each `beforeAll`/`afterAll` to behave differently, they just register a single `beforeAll`/`afterAll` at the root of the test suite.

So even if you had what you described, the root hooks would be like so:
```
beforeAll # registered by root hook
afterAll # registered by root hook
describe
  beforeAll
  afterAll
  it
  it
  it
  describe
    beforeAll
    afterAll
    it
    it
```

Comment (satanTime) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: Yep..... now I'm wondering how it works if I used `MockInstance` in the nested `beforeAll`....

Comment (dpraul) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: I believe it would still work fine. The `afterAll` would still make sure it gets cleaned up

Comment (satanTime) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: in a case of manual reset - yes. But if I customized a service in the first beforeAll and then customized it again in the nested beforeAll, then this change should disappear once the describe has been finished:

```
beforeAll # registered by root hook
afterAll # registered by root hook
describe
  beforeAll
  afterAll
  it -- 1st customization
  describe
    beforeAll
    afterAll
    it -- 2nd customization
  it -- 2nd is removed and 1st is present
```

Comment (satanTime) [reply libs/ng-mocks/src/lib/common/ng-mocks-stack.ts:147]: Anyway it doesn't affect your PR and adds one more point to the solution with a manual control of the stack state which will be implemented as a separate PR.

**Comment by satanTime** [comment]
Hi @ike18t,

might you take a look why this PR hasn't triggered CircleCI? 
I've checked CircleCI setting, they look right, maybe something in the repo settings here?

**Review by satanTime — APPROVED** [review]

**Comment by satanTime** [comment]
[v12.4.0](https://github.com/ike18t/ng-mocks/releases/tag/v12.4.0) has been released and contains a fix for the issue. Feel free to reopen the issue or to submit a new one if you meet any problems.

**Pull request merged**